### PR TITLE
Claim license

### PR DESCRIPTION
**Pull request by Holymir**
License contracts, unit tests, deploy scripts, one new function in Registry and small naming fixes

### CODE REVIEW

File: contracts/License/IPClaimsLicense.sol [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+import "./IIPClaimsLicense.sol";
+import "../ECRecover/ECTools.sol";
+import "./../IPClaim/IPClaim.sol";
+
+
+contract IPClaimsLicense is IIPClaimsLicense, OwnableUpgradableImplementation {
+using SafeMath for uint256;
+
+    uint256 public licenseFeeInUSD;
+    uint256 public licenseFeeInTokens;
```
Comment (vladichhh) [inline contracts/License/IPClaimsLicense.sol]: do we need explicitly to store fees both in USD & tokens ?

File: contracts/License/IPClaimsLicense.sol [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    function setFeeReceiver(address _feeReceiver) public onlyOwner onlyValidAddress(_feeReceiver) {
+        feeReceiver = _feeReceiver;
+    }
+
+    function getFeeReceiver() public view returns(address) {  
+        return feeReceiver;
+    }
+
+    function setIPTokenAddress(address _token) public onlyOwner onlyValidAddress(_token) {
+        erc20Instance = _token;
```
Comment (vladichhh) [inline contracts/License/IPClaimsLicense.sol]: keep consistency between variable names & setter/getter names

File: contracts/TestContracts/License/IPClaimsLicenseTest.sol [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        testAddress = _testAddress;
+    }
+
+    function getTestAddress() public view returns(address) {
+        return testAddress;
+    }
+/**
+*
+ */
+    function registerLicenseWithEthers(address _ipClaimOwner, address _ipCLaimAddress, uint256 licensePriceInEthers, uint256 _days, bytes signedDataByOwner) public payable
```
Comment (vladichhh) [inline contracts/TestContracts/License/IPClaimsLicenseTest.sol]: add some comments on function level and within the function as well

File: contracts/TestContracts/License/IPClaimsLicenseTest.sol [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+        feeReceiver.transfer(licenseFeeInUSD);
+        messageSigner.transfer(licensePriceInEthers);
+
+        hashToDate[_ipCLaimAddress][msg.sender] = (_days.mul(1 days)).add(now);
+
+        emit RegisterLicense(_ipClaimOwner, msg.sender, _ipCLaimAddress, (_days.mul(1 days)).add(now));
+    }
+
+    function registerLicenseWithTokens(address _ipClaimOwner, address _ipCLaimAddress, uint256 licensePriceInTokens, uint256 _days, bytes signedDataByOwner) public
```
Comment (vladichhh) [inline contracts/TestContracts/License/IPClaimsLicenseTest.sol]: add some comments on function level and within the function as well

File: contracts/License/IPClaimsLicense.sol [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        return erc20Instance;
+    }
+
+    function recover(bytes32 hash, bytes signedDataByOwner) internal pure returns(address) {
+        return ECTools.prefixedRecover(hash, signedDataByOwner);
+    }
+/**
+*
+ */
+    function registerLicenseWithEthers(address _ipClaimOwner, address _ipCLaimAddress, uint256 licensePriceInEthers, uint256 _days, bytes signedDataByOwner) public payable
```
Comment (vladichhh) [inline contracts/License/IPClaimsLicense.sol]: add some comments on function level and within the function as well

File: contracts/License/IPClaimsLicense.sol [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+        feeReceiver.transfer(licenseFeeInUSD);
+        messageSigner.transfer(licensePriceInEthers);
+
+        hashToDate[_ipCLaimAddress][msg.sender] = (_days.mul(1 days)).add(now);
+
+        emit RegisterLicense(_ipClaimOwner, msg.sender, _ipCLaimAddress, (_days.mul(1 days)).add(now));
+    }
+
+    function registerLicenseWithTokens(address _ipClaimOwner, address _ipCLaimAddress, uint256 licensePriceInTokens, uint256 _days, bytes signedDataByOwner) public
```
Comment (vladichhh) [inline contracts/License/IPClaimsLicense.sol]: add some comments on function level and within the function as well

File: contracts/License/IPClaimsLicense.sol [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+/**
+*
+ */
+    function registerLicenseWithEthers(address _ipClaimOwner, address _ipCLaimAddress, uint256 licensePriceInEthers, uint256 _days, bytes signedDataByOwner) public payable
+    onlyValidAddress(_ipClaimOwner)
+    onlyValidAddress(_ipCLaimAddress) {
+    
+        // TODO: calculate usd to wei -> (Oracle)
+        require(msg.value >= licenseFeeInUSD.add(licensePriceInEthers));
+        require(_days <= 366 && _days >= 88);
```
Comment (vladichhh) [inline contracts/License/IPClaimsLicense.sol]: min days could be 89

File: contracts/License/IPClaimsLicense.sol [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        require(hashToDate[_ipCLaimAddress][msg.sender] <= now);
+
+        bytes32 hashOfLicense = keccak256(abi.encodePacked(_ipClaimOwner, msg.sender, _ipCLaimAddress, licensePriceInTokens, _days));
+        address messageSigner = recover(hashOfLicense, signedDataByOwner);
+
+        require(IPClaim(_ipCLaimAddress).owner() == messageSigner);
+        require(_ipClaimOwner == messageSigner);
+
+        assert(ERC20(erc20Instance).transferFrom(msg.sender, feeReceiver, licenseFeeInTokens));
+        assert(ERC20(erc20Instance).transferFrom(msg.sender, feeReceiver, licenseFeeInTokens));
```
Comment (vladichhh) [inline contracts/License/IPClaimsLicense.sol]: two lines are equal
add licensing commission percent

File: contracts/License/IIPClaimsLicense.sol [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    function setLicenseFeeInUSD(uint256 _feeInEthers) public;
+    function getLicenseFeeInUSD() public view returns(uint256);
+    function setLicenseFeeInTokens(uint256 _feeInTokens) public;
+    function getLicenseFeeInTokens() public view returns(uint256);
+    function setFeeReceiver(address _feeReceiver) public;
+    function getFeeReceiver() public view returns(address);
+    function setIPTokenAddress(address _token) public;
+    function getIPTokenAddress() public view returns(address);
+    function recover(bytes32 hash, bytes signedDataByOwner) internal pure returns(address);
+    function registerLicenseWithEthers(address _ipClaimOwner, address _ipCLaimAddress, uint256 licensePriceInEthers, uint256 _days, bytes signedDataByOwner) public payable;
```
Comment (vladichhh) [inline contracts/License/IIPClaimsLicense.sol]: its better to be external

File: contracts/License/IIPClaimsLicense.sol [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+    function registerLicenseWithTokens(address _ipClaimOwner, address _ipCLaimAddress, uint256 licensePriceInTokens, uint256 _days, bytes signedDataByOwner) public;
```
Comment (vladichhh) [inline contracts/License/IIPClaimsLicense.sol]: its better to be external

File: contracts/License/IPClaimsLicense.sol [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+* @param _ipCLaimAddress - address of the IPCLaim contract
+* @param licensePriceInTokens - price that licensee should pay to claim owner in tokens
+* @param _days - period of the license in days
+* @param signedDataByOwner - hash of the signed data /claim owner address, licensee address, IPClaim address, price in tokens, days/
+ */
+    function registerLicenseWithTokens(address _ipClaimOwner, address _ipCLaimAddress, uint256 licensePriceInTokens, uint256 _days, bytes signedDataByOwner) external 
+    onlyValidAddress(_ipClaimOwner)
+    onlyValidAddress(_ipCLaimAddress) {
+    
+        require(_days <= 365 && _days >= 88);
```
Comment (vladichhh) [inline contracts/License/IPClaimsLicense.sol]: min amount of days should be 89

File: contracts/License/IPClaimsLicense.sol [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+
+    function registerLicenseWithEthers(address _ipClaimOwner,
+        address _ipCLaimAddress,
+        uint256 licensePriceInEthers,
+        uint256 _days,
+        bytes signedDataByOwner) external payable
+    onlyValidAddress(_ipClaimOwner)
+    onlyValidAddress(_ipCLaimAddress)
+    onlyWithinValidPeriod(_days)
+    onlyNotActiveLicense(_ipCLaimAddress) {
```
Comment (thcrnk) [inline contracts/License/IPClaimsLicense.sol]: Do we need an option for our users to extend license. In current case they could not extend it and could have a few days without license in which legally they do not have permission to use the IP. Do you think we could add extending easy?

File: contracts/License/IPClaimsLicense.sol [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+import "../Upgradability/OwnableUpgradableImplementation/OwnableUpgradableImplementation.sol";
+import "../../node_modules/openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "../TestContracts/Token/ERC20.sol";
+import "./IIPClaimsLicense.sol";
+import "../ECRecover/ECTools.sol";
+import "./../IPClaim/IPClaim.sol";
+import "../Oracle/ExchangeOracle.sol";
+
+
+contract IPClaimsLicense is IIPClaimsLicense, OwnableUpgradableImplementation {
```
Comment (thcrnk) [inline contracts/License/IPClaimsLicense.sol]: Try to add string error messages on all requires.

File: .etherlime-store/.history.json [hunk 14]
```diff
@@ -1 +1 @@
...14676ae9f3085860DC14D918ba1D21094"},{"eventTimestamp":1534338157913,"deployerType":"EtherlimeGanacheDeployer","nameOrLabel":"VaultitudeUpgradableImplsProxy","transactionHash":"0xc95a4360387bbdb701973d64fc2c0857ab5dd8a73e2d5aea65042022ae3b3c7a","status":0,"gasPrice":"20000000000","gasUsed":"164961","result":"0x6e01988B18799140eb8D8F7e1d21b3FABD01959b"},{"eventTimestamp":1534338157968,"deployerType":"EtherlimeGanacheDeployer","nameOrLabel":"ECTools","transactionHash":"0x71395556fa08e780ec41b391fe96a232077c131473d89a32754d75a00000b308","status":0,"gasPrice":"20000000000","gasUsed":"354808","result":"0x7FE6d14d9D418bfA08683714d77BAEB23932C196"},{"eventTimestamp":1534338158048,"deployerType":"EtherlimeGanacheDeployer","nameOrLabel":"IPClaimsTransfer","transactionHash":"0x968ea3574878df1d4081a4d60cdba3185c5731483ea6735f71db98db8fd01186","status":0,"gasPrice":"20000000000","gasUsed":"2558962","result":"0x587AeD2cC47FEf698Ab821af0b8D3831577Fbb0E"},{"eventTimestamp":1534338158111,"deployerType":"EtherlimeGanacheDeployer","nameOrLabel":"IPClaimsTransferProxy","transactionHash":"0xc63fde9cf9bfe51d1d8c4133aafc58ade194b8458f7885c83752d5623ce5f058","status":0,"gasPrice":"20000000000","gasUsed":"
```
Comment (vladichhh) [inline .etherlime-store/.history.json]: do we need to store this .history.json ?

File: contracts/License/IIPClaimsLicense.sol [hunk 15]
```diff
@@ -0,0 +1,30 @@
+pragma solidity ^0.4.23;
+
+import "../Upgradability/OwnableUpgradableImplementation/IOwnableUpgradableImplementation.sol";
+
+
+contract IIPClaimsLicense is IOwnableUpgradableImplementation {
+
+    event RegisterLicense(address indexed claimOwner, address indexed ipClaimLicenseReceiver, address indexed claimAddress, uint256 endDate);
```
Comment (vladichhh) [inline contracts/License/IIPClaimsLicense.sol]: stick to the naming rule for events: nounVerb

File: contracts/Registry/IPClaimsRegistry.sol [hunk 16]
```diff
@@ -57,10 +57,14 @@ contract IPClaimsRegistry is IIPClaimsRegistry, OwnableUpgradableImplementation
     }
 
     function getClaimersCount() public view returns (uint) {
-        return claimers.length;
+        return addressesOfIPClaimers.length;
     }
 
-    function getListOfIPClaims(address claimOwner) public view returns (address) {
+    function getClaimerAddress(uint _arrayPosition) public view returns (address) {
```
Comment (vladichhh) [inline contracts/Registry/IPClaimsRegistry.sol]: check all places and adjust uint to uin256

**Pull request merged**